### PR TITLE
Add ContainerStateExited and OCI delete() in cleanup()

### PR DESCRIPTION
**Pull request by mheon**
To work better with Kata containers, we need to delete() from the OCI runtime as a part of cleanup, to ensure resources aren't retained longer than they need to be.

To enable this, we need to add a new state to containers, ContainerStateExited. Containers transition from ContainerStateStopped to ContainerStateExited via cleanupRuntime which is invoked as part of cleanup(). A container in the Exited state is identical to Stopped, except it has been removed from the OCI runtime and thus will be handled differently when initializing the container.

Same goal as #1498 but should handle container restart properly.

### CODE REVIEW

File: libpod/container_internal.go [hunk 1]
```diff
@@ -845,6 +846,15 @@ func (c *Container) cleanup() error {
 		}
 	}
 
+	// Remove the container from the runtime, if necessary
+	if err := c.cleanupRuntime(ctx); err != nil {
+		if lastError != nil {
+			logrus.Errorf("Error removing container %s from OCI runtime: %v", c.ID(), err)
+		} else {
```
Comment (rhatdan) [inline libpod/container_internal.go]: This should not be an else here.  You wan to always set lastError=err.

File: libpod/container_internal.go [hunk 2]
```diff
@@ -150,7 +150,8 @@ func (c *Container) syncContainer() error {
 	// If runtime knows about the container, update its status in runtime
 	// And then save back to disk
 	if (c.state.State != ContainerStateUnknown) &&
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal.go]: Might be good to create a "isContainerReady()"  type of function at some point so if we need to adjust down the line we can just change the one function.  Might be a little more readable too.  Not a stopper for this PR, just a suggestion.

File: libpod/container_internal.go [hunk 3]
```diff
@@ -552,19 +558,30 @@ func (c *Container) reinit(ctx context.Context) error {
...
+// Should only be done on ContainerStateStopped containers.
+// Not necessary for ContainerStateExited - the container has already been
+// removed from the runtime, so init() can proceed freely.
+func (c *Container) reinit(ctx context.Context) error {
+	logrus.Debugf("Recreating container %s in OCI runtime", c.ID())
+
+	if err := c.cleanupRuntime(ctx); err != nil {
+		return err
+	}
+
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal.go]: In the comments you say this should only be done on stopped containers but I don't see a state check here.  Do you need one?

File: libpod/container_internal.go [hunk 4]
```diff
@@ -603,28 +620,11 @@ func (c *Container) initAndStart(ctx context.Context) (err error) {
...
-		if err := c.save(); err != nil {
+		if err := c.reinit(ctx); err != nil {
 			return err
 		}
-	}
-
-	// If we are ContainerStateConfigured we need to init()
-	if c.state.State == ContainerStateConfigured {
+	} else if c.state.State == ContainerStateConfigured ||
+		c.state.State == ContainerStateExited {
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal.go]: Should Unknown state be checked here too?

File: libpod/container_internal.go [hunk 5]
```diff
@@ -716,8 +716,9 @@ func (c *Container) restartWithTimeout(ctx context.Context, timeout uint) (err e
 		if err := c.reinit(ctx); err != nil {
 			return err
 		}
-	} else if c.state.State == ContainerStateConfigured {
-		// Initialize the container if it has never been initialized
+	} else if c.state.State == ContainerStateConfigured ||
+		c.state.State == ContainerStateExited {
+		// Initialize the container
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal.go]: Ditto unknown.

File: libpod/container_internal.go (hunk 1 above)
Comment (mheon) [reply libpod/container_internal.go]: The way this one is set up, we print the new error, and return lastError. I can rework it if you want, but I'd also have to change the other one here.

File: libpod/container_internal.go (hunk 4 above)
Comment (mheon) [reply libpod/container_internal.go]: ContainerStateUnknown is a hard error in syncContainer right after we begin the function, so we should never get here if we are unknown

File: libpod/container_internal.go (hunk 1 above)
Comment (rhatdan) [reply libpod/container_internal.go]: Ok I missed that.  So you want to return the first error that happens. SGTM

File: libpod/container_internal.go (hunk 2 above)
Comment (rhatdan) [reply libpod/container_internal.go]: I was thinking the same thing, since we have this multi line if in many places.

**Comment by mheon** [comment]
Network flakes out of Travis

**Comment by mheon** [comment]
bot, retest this please

**Comment by mheon** [comment]
Tests finally green
@rhatdan @TomSweeneyRedHat @baude @umohnani8 @vrothberg PTAL

**Comment by baude** [comment]
LGTM, were previous comments all addressed ?

**Comment by mheon** [comment]
I don't think we can do an isContainerReady() easily given the varying definition of what ready is for various API calls. Maybe worth looking into, but in a separate PR.

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Pull request merged**